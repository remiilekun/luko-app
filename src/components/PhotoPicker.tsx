import {
  Image,
  ImageStyle,
  Pressable,
  StyleSheet,
  Text,
  TextStyle,
  View,
  ViewStyle,
} from "react-native";
import { Ionicons, Entypo } from "@expo/vector-icons";
import { colors } from "../theme/colors";
import useImagePicker, { ImagePickerResult } from "../hooks/useImagePicker";
import { fonts } from "../theme/fonts";

interface PhotoPickerProps {
  error?: string;
  onChange?: (value: string) => void;
  photo?: string;
}

const PhotoPicker = ({ error, photo, onChange }: PhotoPickerProps) => {
  const { launchPicker, picker } = useImagePicker({ onChange });

  const onDelete = () => {
    onChange?.("");
  };

  return (
    <>
      {photo ? (
        <View style={styles.pickerContainer}>
          <Image
            resizeMode="cover"
            source={{ uri: photo }}
            style={styles.previewImage}
          />
          <Pressable style={styles.deleteWrapper} onPress={onDelete}>
            <Ionicons name="md-trash" size={15} color={colors.white} />
          </Pressable>
        </View>
      ) : (
        <Pressable style={styles.pickerContainer} onPress={launchPicker}>
          <Entypo
            color={colors.mainBlue}
            name="camera"
            size={44}
            style={styles.pickerIcon}
          />
          <Text style={styles.pickerText}>Add photo</Text>
        </Pressable>
      )}

      {error ? <Text style={styles.errorText}>{error}</Text> : null}
      {picker}
    </>
  );
};

export default PhotoPicker;
export { ImagePickerResult };

const styles = StyleSheet.create({
  deleteWrapper: {
    alignItems: "center",
    backgroundColor: colors.red,
    borderRadius: 16,
    bottom: 0,
    height: 32,
    justifyContent: "center",
    position: "absolute",
    right: 0,
    width: 32,
  },
  previewImage: {
    borderRadius: 75,
    height: 150,
    width: 150,
  } as ImageStyle,
  pickerContainer: {
    alignItems: "center",
    // lightGrey not clear here, replacing with mainGrey
    borderColor: colors.mainGrey,
    borderRadius: 75,
    borderStyle: "dashed",
    borderWidth: 2,
    height: 150,
    justifyContent: "center",
    position: "relative",
    width: 150,
  } as ViewStyle,
  pickerIcon: {
    marginBottom: 14,
  } as ViewStyle,
  pickerText: {
    color: colors.black,
    fontSize: 17,
    fontFamily: fonts.regular,
  } as TextStyle,
  errorText: {
    color: colors.red,
    fontFamily: fonts.regular,
    fontSize: 10,
    fontWeight: "400",
    marginTop: 5,
  } as TextStyle,
});
