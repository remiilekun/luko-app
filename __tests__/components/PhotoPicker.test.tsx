import "react-native";
import React from "react";
import { fireEvent, render } from "@testing-library/react-native";
import PhotoPicker from "../../src/components/PhotoPicker";

describe("components", () => {
  describe("PhotoPicker", () => {
    it("renders correctly", async () => {
      const { toJSON } = render(<PhotoPicker />);
      expect(toJSON()).toMatchSnapshot();
    });

    it("renders picker if photo not available", async () => {
      const { getByTestId } = render(<PhotoPicker />);
      const selectButton = getByTestId("select-button");
      expect(selectButton).not.toBeNull();
    });

    it("renders photo if available", async () => {
      const { getByTestId } = render(
        <PhotoPicker photo="https://picsum.photos/200" />
      );

      const previewImage = await getByTestId("preview-image");
      expect(previewImage).not.toBeNull();
    });

    it("renders delete button if photo is available", async () => {
      const onChange = jest.fn();
      const { getByTestId } = render(
        <PhotoPicker photo="https://picsum.photos/200" onChange={onChange} />
      );

      const deleteButton = getByTestId("delete-button");
      expect(deleteButton).not.toBeNull();
      fireEvent(deleteButton, "press");
      expect(onChange).toHaveBeenCalledWith("");
    });
  });
});
