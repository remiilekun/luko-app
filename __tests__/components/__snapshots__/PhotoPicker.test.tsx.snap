// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components PhotoPicker renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderColor": "#C0BEB8",
      "borderRadius": 75,
      "borderStyle": "dashed",
      "borderWidth": 2,
      "height": 150,
      "justifyContent": "center",
      "position": "relative",
      "width": 150,
    }
  }
  testID="select-button"
>
  <Text />
  <Text
    style={
      Object {
        "color": "#000000",
        "fontFamily": "Circular-Regular",
        "fontSize": 17,
      }
    }
  >
    Add photo
  </Text>
</View>
`;
